File: frontend/src/app/services/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export enum TaskStatus {
  TO_DO = 'TO_DO',
  IN_PROGRESS = 'IN_PROGRESS',
  DONE = 'DONE'
}

export interface Task {
  id?: number;
  title: string;
  description?: string;
  status: TaskStatus;
  projectId: number;
  assignedTo?: number;
  assignedToName?: string;
  createdBy: number;
  createdByName?: string;
  dueDate?: string;
  completedAt?: string;
  priority: number; // 0 = Low, 1 = Medium, 2 = High
  orderIndex: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface CreateTaskRequest {
  title: string;
  description?: string;
  projectId: number;
  assignedTo?: number;
  createdBy: number;
  dueDate?: string;
  priority?: number;
  orderIndex?: number;
}

export interface UpdateTaskRequest {
  title?: string;
  description?: string;
  status?: TaskStatus;
  assignedTo?: number;
  dueDate?: string;
  priority?: number;
  orderIndex?: number;
}

export interface TaskStatistics {
  total: number;
  toDo: number;
  inProgress: number;
  done: number;
  completionRate: number;
}

@Injectable({
  providedIn: 'root'
})
export class TaskService {
  private apiUrl = 'http://localhost:8082/api/tasks';

  constructor(private http: HttpClient) {}

  /**
   * Create a new task
   */
  createTask(request: CreateTaskRequest): Observable<Task> {
    return this.http.post<Task>(this.apiUrl, request);
  }

  /**
   * Get all tasks for a project
   */
  getTasksByProject(projectId: number): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.apiUrl}/project/${projectId}`);
  }

  /**
   * Get tasks by project and status
   */
  getTasksByProjectAndStatus(projectId: number, status: TaskStatus): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.apiUrl}/project/${projectId}/status/${status}`);
  }

  /**
   * Get task by ID
   */
  getTaskById(id: number): Observable<Task> {
    return this.http.get<Task>(`${this.apiUrl}/${id}`);
  }

  /**
   * Update a task
   */
  updateTask(id: number, request: UpdateTaskRequest): Observable<Task> {
    return this.http.put<Task>(`${this.apiUrl}/${id}`, request);
  }

  /**
   * Delete a task
   */
  deleteTask(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /**
   * Get task statistics for a project
   */
  getTaskStatistics(projectId: number): Observable<TaskStatistics> {
    return this.http.get<TaskStatistics>(`${this.apiUrl}/project/${projectId}/statistics`);
  }

  /**
   * Get status display name
   */
  getStatusDisplayName(status: TaskStatus): string {
    const statusMap: { [key in TaskStatus]: string } = {
      [TaskStatus.TO_DO]: 'To Do',
      [TaskStatus.IN_PROGRESS]: 'In Progress',
      [TaskStatus.DONE]: 'Done'
    };
    return statusMap[status] || status;
  }

  /**
   * Get priority display name
   */
  getPriorityDisplayName(priority: number): string {
    const priorityMap: { [key: number]: string } = {
      0: 'Low',
      1: 'Medium',
      2: 'High'
    };
    return priorityMap[priority] || 'Low';
  }

  /**
   * Get priority color class
   */
  getPriorityColorClass(priority: number): string {
    const colorMap: { [key: number]: string } = {
      0: 'priority-low',
      1: 'priority-medium',
      2: 'priority-high'
    };
    return colorMap[priority] || 'priority-low';
  }
}

File: services/project-service/src/main/java/com/freelance/projectservice/service/TaskService.java
package com.freelance.projectservice.service;

import com.freelance.projectservice.dto.CreateTaskRequest;
import com.freelance.projectservice.dto.TaskDTO;
import com.freelance.projectservice.dto.UpdateTaskRequest;
import com.freelance.projectservice.exception.ResourceNotFoundException;
import com.freelance.projectservice.model.Task;
import com.freelance.projectservice.model.TaskStatus;
import com.freelance.projectservice.repository.ProjectRepository;
import com.freelance.projectservice.repository.TaskRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.format.DateTimeParseException;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class TaskService {
    
    @Autowired
    private TaskRepository taskRepository;
    
    @Autowired
    private ProjectRepository projectRepository;
    
    /**
     * Create a new task
     */
    public TaskDTO createTask(CreateTaskRequest request) {
        // Verify project exists
        if (!projectRepository.existsById(request.getProjectId())) {
            throw new ResourceNotFoundException("Project not found with id: " + request.getProjectId());
        }
        
        Task task = new Task();
        task.setTitle(request.getTitle());
        task.setDescription(blankToNull(request.getDescription()));
        task.setProjectId(request.getProjectId());
        task.setAssignedTo(request.getAssignedTo());
        task.setCreatedBy(request.getCreatedBy());
        task.setDueDate(parseDueDate(request.getDueDate()));
        task.setPriority(request.getPriority() != null ? request.getPriority() : 0);
        task.setOrderIndex(request.getOrderIndex() != null ? request.getOrderIndex() : 0);
        task.setStatus(TaskStatus.TO_DO);
        
        Task savedTask = taskRepository.save(task);
        return convertToDTO(savedTask);
    }
    
    /**
     * Get all tasks for a project
     */
    @Transactional(readOnly = true)
    public List<TaskDTO> getTasksByProject(Long projectId) {
        List<Task> tasks = taskRepository.findByProjectIdOrderByOrderIndexAsc(projectId);
        return tasks.stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }
    
    /**
     * Get task by ID
     */
    @Transactional(readOnly = true)
    public TaskDTO getTaskById(Long id) {
        Task task = taskRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Task not found with id: " + id));
        return convertToDTO(task);
    }
    
    /**
     * Update a task
     */
    public TaskDTO updateTask(Long id, UpdateTaskRequest request) {
        Task task = taskRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Task not found with id: " + id));
        
        if (request.getTitle() != null) {
            task.setTitle(request.getTitle());
        }
        if (request.getDescription() != null) {
            task.setDescription(request.getDescription());
        }
        if (request.getStatus() != null) {
            task.setStatus(request.getStatus());
        }
        if (request.getAssignedTo() != null) {
            task.setAssignedTo(request.getAssignedTo());
        }
        if (request.getDueDate() != null) {
            task.setDueDate(request.getDueDate());
        }
        if (request.getPriority() != null) {
            task.setPriority(request.getPriority());
        }
        if (request.getOrderIndex() != null) {
            task.setOrderIndex(request.getOrderIndex());
        }
        
        Task updatedTask = taskRepository.save(task);
        return convertToDTO(updatedTask);
    }
    
    /**
     * Delete a task
     */
    public void deleteTask(Long id) {
        if (!taskRepository.existsById(id)) {
            throw new ResourceNotFoundException("Task not found with id: " + id);
        }
        taskRepository.deleteById(id);
    }
    
    /**
     * Get tasks by status for a project
     */
    @Transactional(readOnly = true)
    public List<TaskDTO> getTasksByProjectAndStatus(Long projectId, TaskStatus status) {
        List<Task> tasks = taskRepository.findByProjectIdAndStatusOrderByOrderIndexAsc(projectId, status);
        return tasks.stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }
    
    /**
     * Get task statistics for a project
     */
    @Transactional(readOnly = true)
    public TaskStatistics getTaskStatistics(Long projectId) {
        long total = taskRepository.countByProjectId(projectId);
        long toDo = taskRepository.countByProjectIdAndStatus(projectId, TaskStatus.TO_DO);
        long inProgress = taskRepository.countByProjectIdAndStatus(projectId, TaskStatus.IN_PROGRESS);
        long done = taskRepository.countByProjectIdAndStatus(projectId, TaskStatus.DONE);
        
        return new TaskStatistics(total, toDo, inProgress, done);
    }
    
    private static String blankToNull(String s) {
        return s != null && !s.trim().isEmpty() ? s.trim() : null;
    }

    private static LocalDateTime parseDueDate(String value) {
        if (value == null || value.trim().isEmpty()) return null;
        String v = value.trim().replace("Z", "").replace("z", "");
        if (v.length() > 19) v = v.substring(0, 19);
        try {
            return LocalDateTime.parse(v);
        } catch (DateTimeParseException e) {
            return null;
        }
    }

    /**
     * Convert Task entity to DTO
     */
    private TaskDTO convertToDTO(Task task) {
        TaskDTO dto = new TaskDTO();
        dto.setId(task.getId());
        dto.setTitle(task.getTitle());
        dto.setDescription(task.getDescription());
        dto.setStatus(task.getStatus());
        dto.setProjectId(task.getProjectId());
        dto.setAssignedTo(task.getAssignedTo());
        dto.setCreatedBy(task.getCreatedBy());
        dto.setDueDate(task.getDueDate());
        dto.setCompletedAt(task.getCompletedAt());
        dto.setPriority(task.getPriority());
        dto.setOrderIndex(task.getOrderIndex());
        dto.setCreatedAt(task.getCreatedAt());
        dto.setUpdatedAt(task.getUpdatedAt());
        // Note: assignedToName and createdByName should be populated by calling user-service
        return dto;
    }
    
    /**
     * Task statistics inner class
     */
    public static class TaskStatistics {
        private long total;
        private long toDo;
        private long inProgress;
        private long done;
        
        public TaskStatistics(long total, long toDo, long inProgress, long done) {
            this.total = total;
            this.toDo = toDo;
            this.inProgress = inProgress;
            this.done = done;
        }
        
        // Getters
        public long getTotal() { return total; }
        public long getToDo() { return toDo; }
        public long getInProgress() { return inProgress; }
        public long getDone() { return done; }
        
        public double getCompletionRate() {
            return total > 0 ? (done * 100.0 / total) : 0.0;
        }
    }
}
